source/common/stats/allocator_impl.cc

```diff
@@ -316,5 +336,79 @@ Counter* AllocatorImpl::makeCounterInternal(StatName name, StatName tag_extracte
   return new CounterImpl(name, *this, tag_extracted_name, stat_name_tags);
 }
 
+void AllocatorImpl::forEachCounter(std::function<void(std::size_t)> f_size,
+                                   std::function<void(Stats::Counter&)> f_stat) const {
+  Thread::LockGuard lock(mutex_);
+  f_size(counters_.size());
+  for (auto& counter : counters_) {
+    f_stat(*counter);
+  }
+}
+
+void AllocatorImpl::forEachGauge(std::function<void(std::size_t)> f_size,
+                                 std::function<void(Stats::Gauge&)> f_stat) const {
```

Comment:
it's not super-obvious that everyone who would want to call forEachGauge on an allocator would want to skip the uninitialized ones.

I'm wondering if it makes more sense to just let the caller do this filtering. The main argument for having this be a feature of this API is that you could hold the lock while counting them and then enumerating them.

Maybe let the caller provide the predicate, and offer that as a feature for the other stat types?

I'm OK if you want to do that as a follow-up, but it'd be nice to offer a consistent API eventually.

Reply:
I could see you might want to similarly iterated over counters where used()==true

Reply:
Made this change.

Added the check to gauges().